# Contents

#### 01- Import necessary Libraries
#### 02- Import dataframes orders.csv and products.csv
#### 03- Mixed type data
#### 04- Finding Missing Values
#### 05- Addressing Missing Values
#### 06- Finding Duplicates
#### 07- Addressing Duplicates
### Task 4.5
#### 02- Consistency Check on df_ords
#### 03- Mixed type data
#### 04- Fixing mixed type data
#### 05- Checking Missing Values
#### 06- Addressing Missing Values
#### 07- Checking for Duplicates
#### 08- Addressing Duplicates
#### 09- Exporting dataframes

# 01. Importing Libraries

In [1]:
# Import Libraries 

import pandas as pd
import numpy as np
import os

# 02. Import Dataframes

In [2]:
# Creating path for products.csv and orders.csv

path = r'C:\Users\aleen\OneDrive\Desktop\Aleena\CF assignments\Data Immersion\Achievement 4\23-04-2023 Instacart Basket Analysis'

In [3]:
# Importing products.csv as df_prods

df_prods =  pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Importing wrangled_orders.csv as df_ords

df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'wrangled_orders.csv'), index_col = False)

# 03. Mixed-Type Data

In [5]:
# Create a data frame

df_test = pd.DataFrame()

In [6]:
# Create a mixed type column

df_test['mix'] = ['a', 'b', 1, True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
# Check for mixed column

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [10]:
# Changing the data type

df_test['mix'] = df_test['mix'].astype('str')

# 04. Finding Missing Values

In [11]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# 05. Addressing Missing Values

In [14]:
df_prods.shape

(49693, 5)

In [15]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [16]:
df_prods_clean.shape

(49677, 5)

# 06. Finding Duplicates

In [17]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

# 07. Addressing Duplicates

In [19]:
df_prods_clean.shape

(49677, 5)

In [20]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5

# 02. Consistency Check on df_ords

In [22]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

# 03. Mixed type data

In [24]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
eval_set                   object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# 04. Fixing mixed type data

No mixed data type found.

# 05. Checking missing values

In [25]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

column 'day_since_prior_order' has missing values 206209

# 06. Addressing missing values

In [32]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [33]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

The user_id with order_number 1 has NaN value of days_since_prio_order; which should not be considered as missing value as this is their first order and they won't be having days_since_prior order. 

# 07. Checking for Duplicates

In [34]:
df_ords_dups = df_ords[df_ords.duplicated()]

In [35]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

# 08. Addressing Duplicates

In [36]:
df_ords_dups.size

0

No duplicates found.

# 09. Exporting dataframes

In [37]:
# Exporting checked df_prods

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'checked_products.csv'))

In [38]:
# Exporting checked df_ords

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'checked_orders.csv'))